src/test/fio/fio_ceph_messenger.cc

```diff
@@ -0,0 +1,690 @@
+    }
+    single_msgr_disps.push_back(disp);
+    single_msgr_ref++;
+  } else {
+    /*
+     * Messenger instance per FIO thread
+     */
+    msgr = create_messenger(o);
+  }
+  msgr->add_dispatcher_head(disp);
+
+  data->disp = disp;
+  data->msgr = msgr;
+}
+
+static void free_messenger(struct ceph_msgr_data *data)
+{
+  data->msgr->shutdown();
+  data->msgr->wait();
+  delete data->msgr;
+}
+
+static void put_messenger(struct ceph_msgr_data *data)
+{
+  struct ceph_msgr_options *o = data->o;
+
+  if (o->is_single) {
+    if (--single_msgr_ref == 0) {
+      free_messenger(data);
+      /*
+       * In case of a single messenger instance we have to
+       * free dispatchers after actual messenger destruction.
+       */
+      for (auto disp : single_msgr_disps)
+	delete disp;
+      single_msgr = NULL;
+    }
+  } else {
+    free_messenger(data);
+    delete data->disp;
+  }
+  data->disp = NULL;
+  data->msgr = NULL;
+}
+
+static int fio_ceph_msgr_setup(struct thread_data *td)
+{
+  struct ceph_msgr_options *o = (decltype(o))td->eo;
+  ceph_msgr_data *data;
+
+  /* We have to manage global resources so we use threads */
+  td->o.use_thread = 1;
+
+  create_or_get_ceph_context(o);
+
+  if (!td->io_ops_data) {
+    data = new ceph_msgr_data(o, td->o.iodepth);
+    init_messenger(data);
+    td->io_ops_data = (void *)data;
+  }
+
+  return 0;
+}
+
+static void fio_ceph_msgr_cleanup(struct thread_data *td)
+{
+  struct ceph_msgr_data *data;
+  unsigned nr;
+
+  data = (decltype(data))td->io_ops_data;
+  put_messenger(data);
+
+  nr = ring_buffer_used_size(&data->io_completed_q);
+  if (nr)
+    fprintf(stderr, "fio: io_completed_nr==%d, but should be zero\n",
+	    nr);
+  if (data->io_inflight_nr)
+    fprintf(stderr, "fio: io_inflight_nr==%d, but should be zero\n",
+	    data->io_inflight_nr);
+  if (data->io_pending_nr)
+    fprintf(stderr, "fio: io_pending_nr==%d, but should be zero\n",
+	    data->io_pending_nr);
+  if (!flist_empty(&data->io_inflight_list))
+    fprintf(stderr, "fio: io_inflight_list is not empty\n");
+  if (!flist_empty(&data->io_pending_list))
+    fprintf(stderr, "fio: io_pending_list is not empty\n");
+
+  ring_buffer_deinit(&data->io_completed_q);
+  delete data;
+  put_ceph_context();
+}
+
+static int fio_ceph_msgr_io_u_init(struct thread_data *td, struct io_u *io_u)
+{
+  struct ceph_msgr_options *o = (decltype(o))td->eo;
+  struct ceph_msgr_io *io;
+  MOSDOp *req_msg = NULL;
+
+  io = (decltype(io))malloc(sizeof(*io));
+  io->io_u = io_u;
+  io->data = (decltype(io->data))td->io_ops_data;
+
+  if (!o->is_receiver) {
+    object_t oid(ptr_to_str(io));
+    pg_t pgid;
+    object_locator_t oloc;
+    hobject_t hobj(oid, oloc.key, CEPH_NOSNAP, pgid.ps(),
+		   pgid.pool(), oloc.nspace);
+    spg_t spgid(pgid);
+    entity_inst_t dest(entity_name_t::OSD(0), hostname_to_addr(o));
+
+    req_msg = new MOSDOp(0, 0, hobj, spgid, 0, 0, 0);
+    req_msg->set_connection(io->data->msgr->get_connection(dest));
```

Comment:
get_connection() was removed recently hence build is failing...

Reply:
Yep, will update shortly.

Reply:
Please check.